### PR TITLE
[16.0][IMP] sale_order_product_picker: Support use of discounts

### LINKED ISSUE #2842: sale_order_product_picker: quick-clicking on +1 button fails
For example, if you want to add 3 units and click the button 3 times quickly, chances are that only 1 or 2 of those clicks are actually handled.

I think it would make sense to throttle and defer the calls, and only when the user has stopped clicking for about 500ms, then upload the changes. Does that make sense?

Otherwise, at least it would help if the button appears as disabled while you can't use it. That visual feedback isn't the best, but at least gives a hint that you have to wait to click more.

_Originally posted by @yajo in https://github.com/OCA/sale-workflow/pull/2796#pullrequestreview-1743843628_

### PR DESCRIPTION
**Pull request by CarlosRoca13**
cc @Tecnativa TT48323

ping @sergio-teruel @carlosdauden

### CODE REVIEW

File: sale_order_product_picker/models/sale_order_picker.py [hunk 1]
```diff
@@ -100,21 +287,32 @@ def _get_last_sale_price_product(self, use_delivery_address=False):
...
             .with_context(prefetch_fields=False)
         )
-        return so_line.price_unit or 0.0
+        return so_line.price_unit, so_line.discount, so_line.price_reduce
 
     def add_to_cart(self):
         self.ensure_one()
         so_line = self.order_id.order_line.new({"product_id": self.product_id.id})
-        so_line.product_id_change()
+        # so_line._product_id_change()
```
Comment (sergio-teruel) [inline sale_order_product_picker/models/sale_order_picker.py]: Is ok this commented line??

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

File: sale_order_product_picker/models/sale_order_picker.py (hunk 1 above)
Comment (CarlosRoca13) [reply sale_order_product_picker/models/sale_order_picker.py]: The method doesn't exist now, removed line

File: sale_order_product_picker/static/src/x2many/x2many_field.esm.js [hunk 2]
```diff
@@ -45,16 +72,21 @@ patch(X2ManyField.prototype, "sale_order_product_picker.X2ManyField", {
...
+                        line_price_reduce: price_discount[1],
                     });
                 }
                 return res;
             };
-            const newUpdateRecord = (record) => {
+            const newUpdateRecord = async (record) => {
                 // Function used when updating a record.
-                const res = updateRecord(record);
+                lineList = this.props.record.model.root.data.order_line;
```
Comment (chienandalu) [inline sale_order_product_picker/static/src/x2many/x2many_field.esm.js:85]: You aren't declaring it... doesn't this fail?

File: sale_order_product_picker/static/src/x2many/x2many_field.esm.js [hunk 3]
```diff
@@ -67,28 +99,41 @@ patch(X2ManyField.prototype, "sale_order_product_picker.X2ManyField", {
...
+                        line_price_reduce: price_discount[1],
                     });
                 }
                 return res;
             };
-            const newRemoveRecord = (record) => {
+            const newRemoveRecord = async (record) => {
                 // Function used when removing a record.
-                const res = removeRecord(record);
+                lineList = this.props.record.model.root.data.order_line;
```
Comment (chienandalu) [inline sale_order_product_picker/static/src/x2many/x2many_field.esm.js:115]: Same here..

**Comment by CarlosRoca13** [comment]
That var is defined in line 29. I added comments explaining why need to redefine the list.

**Comment by CarlosRoca13** [comment]
Added changes to make the button +1 faster. With these changes, what we do is to store the changes to be made while pressing the +1 button. Each time the button is pressed, the timer that will execute the writing of the lines is reset. 

This relieves the server of unnecessary calls, since everything will be written at the same time. And the user will not have the feeling of slowness when writing on 3G networks.

Here is a sample:
Example picker16 0

As you can see in the example, the time to write the pending lines is 1 sec, but it can be increased using the system parameter `sale_order_product_picker.delay` and the number of seconds to wait.

**Comment by CarlosRoca13** [comment]
@yajo I believe this should resolve this issue :smile: https://github.com/OCA/sale-workflow/issues/2842

**Comment by sergio-teruel** [comment]
@CarlosRoca13 Can you put this PR as draft until september?

**Comment by CarlosRoca13** [comment]
This PR is ready for review

**Review by yajo — APPROVED** [review]
Thanks!

**Review by sergio-teruel — APPROVED** [review]

**Pull request merged**